torch/utils/bundled_inputs.py

```diff
@@ -0,0 +1,148 @@
+    model._c._register_attribute("_bundled_inputs_deflated", deflated_inputs_type, [])
+    model._c._register_attribute("_bundled_inputs_inflated", inflated_inputs_type, None)
+
+    if hasattr(model, "_generate_bundled_inputs"):
+        if inputs is not None:
+            raise Exception(
+                "inputs is not None, but _generate_bundled_inputs is already defined")
+        # Model author already defined _generate_bundled_inputs.
+    elif inputs is None:
+        raise Exception(
+            "inputs must be specified if _generate_bundled_inputs is not already defined")
+    else:
+        # Iterate over the inputs and args in each input.
+        # Accumulate `deflated_inputs` as (possibly) compressed values
+        # and `parts` to be joined into the expression that unpacks them.
+        deflated_inputs = []
+        parts = []
+        for inp_idx, args in enumerate(inputs):
+            deflated_args = []
+            parts.append("(")
+            for arg_idx, arg in enumerate(args):
+                deflated, inflater = _inflate_expr(arg, f"deflated[{inp_idx}][{arg_idx}]")
+                deflated_args.append(deflated)
+                parts.append(f"    {inflater},")
+            deflated_inputs.append(tuple(deflated_args))
+            parts.append("),")
+        parts.append("")
+        expr = "\n".join(parts)
+        # Back-channel return this expr for debugging.
+        if _receive_inflate_expr is not None:
+            _receive_inflate_expr.append(expr)
+        model._bundled_inputs_deflated = deflated_inputs
+        definition = textwrap.dedent("""
+            def _generate_bundled_inputs(self):
+                deflated = self._bundled_inputs_deflated
+                return [
+            {}
+                ]
+            """).format(expr)
+        model.define(definition)
+
+    # Define get_all_bundled_inputs that caches the generated inputs.
+    model.define(textwrap.dedent("""
+        def get_all_bundled_inputs(self):
+            if self._bundled_inputs_inflated is None:
+                self._bundled_inputs_inflated = self._generate_bundled_inputs()
+            all_inputs = self._bundled_inputs_inflated
+            assert all_inputs is not None
+            return all_inputs
+        """))
+
+    # Define some helper methods.
+    model.define(textwrap.dedent("""
+        def get_num_bundled_inputs(self):
+            return len(self.get_all_bundled_inputs())
+        """))
+    model.define(textwrap.dedent("""
+        def run_on_bundled_input(self, idx: int):
+            return self(*self.get_all_bundled_inputs()[idx])
+        """))
+
+
+def _inflate_expr(arg: T, ref: str) -> T:
+    if isinstance(arg, torch.Tensor):
+        # Allow custom inflation expressions any tensor.
+        # For example, calling custom image-decoding ops.
+        # Or just use "{}" as the format string to ignore size limits.
+        if hasattr(arg, "_bundled_input_inflate_format"):
```

Comment:
wait, you're suggesting adding attribute to a tensor? it's a bit hacky. Maybe allow input argument to be a callable and if so - invoke it?

Reply:
Yes, it is a bit hacky.  Assuming anything callable is meant to be called seems error-prone.  If you want something cleaner, I think I'd want to go with a NamedTuple defined in this module that has two fields: deflated value and inflation format string.  Then we can just do an isinstance check.

Reply:
yeah, named tuple is probably better, current one is just too magical